### ISSUE TITLE
Suggestion: configure store URL in UI

### ISSUE THREAD

**Issue by s8sg** [openfaas/faas#612]
## Expected Behaviour
Load a store with provided functions definitions from a given link targeted to `store.json`or load a function directly from a git repo that includes `stack.yml`

## Current Behaviour
Currently the store is being loaded from fixed file hosted at
```
https://raw.githubusercontent.com/openfaas/store/master/store.json
```
The functions in store are fixed and can be modified only after modifying the above file

## Possible Solution

The current store items could be loaded and displayed under `default` section in store

**Load store**  
Accepts a url for a `store.json` file in the URL
```
https://faashostedfunction.com/imageprocessing/store.json
```
which will load the function from the given store file, and display them in the list of available functions under the `imageprocessing` section

**Update:** Currently fass-cli provides the feature to query and deploy a store by store location. Although as the context is saved locally the functions wont be available to redeploy until the same URL has been provided. If the Store definition  loaded (saved) in the gateway, it can be reused for deploying function from API, UI or Faas-cli.  

**Load function**  
  
Or a function could be added locally by providing an URL link to function repo
```
"https://github.com/jockdarock/nslookup_faas"
```
once added that function will be displayed on the list under the `other` section

## Context
Currently the store is fixed json file. Having a ability to add store by a url would allow user to load a specific store and use its functions. It might also make the community to come up with stores focused on different domains of problems. Later separation of store by URL would allow to manage different namespace. Like: 
```
https://faasinationgfunction.com/imageprocessing/store.json
https://faasinationgfunction.com/networkfunc/store.json
https://faasinationgfunction.com/machinelearningproblems/store.json
```

**Comment by iyovcheva** [comment]
Derek add label: proposal

**Comment by rgee0** [comment]
@s8sg Please have a look at the [CLI](https://github.com/openfaas/faas-cli) as this facility is already available there.

**Comment by s8sg** [comment]
@rgee0 thanks a lot for pointing out. So as per my understanding the List of functions in Store currently only browsing the provided store in local context of `faas-cli`. The request is not being made to gateway until the function is being deployed. So every time the `fass-cli store` command is fired we need to provide the `url` of the store. It would be nice to have a functionality to save a store definition by the store namespace in the gateway. So that once a store is verified and loaded, the same store can be deployed or queried for function list from any context using: `API, UI or Faas-cli`. A map of store namespace and store definition or http location can be maintained
```bash
imageprocessing/converter   --> https://myfaasstore.com/imageprocessing/converter/store.json
imageprocessing/query  --> https://myfaasstore.com/imageprocessing/query/store.json
```
The command to load a store definition might look like:
```bash
fass-cli store load <store.name> --url <store.url>
```

For now it would be nice to have the store deploy functionalities in the UI as well

**Comment by ericstoekl** [comment]
This should really be moved to the `faas-cli` repo, right? Maybe close this and re-open it there.

**Comment by s8sg** [comment]
@ericstoekl this issue mainly describe the idea of having store handlers in the API provided by the`Gateway` of faas. And then providing the support to `faas-cli` and in the `UI`. I think it is more applicable to `faas` but I'm okay to move it in `faas-cli`. Please confirm

**Comment by ericstoekl** [comment]
I prefer to keep things like the store as very simple template files that you load with only `faas-cli`. The Gateway is for core functionality, and its code base shouldn't be burdened with code unless there is a strong need. @alexellis please weigh in on whether or not the Gateway should be storing Store URLs

**Comment by s8sg** [comment]
@ericstoekl I understand but in my use-case I would need a automated processed to load store definition from fass-cli. Although the list of available function is being acquired using the UI for now. 
As currently both UI and CLI resolves the store.json file locally to generate the list; I can't Use the UI / API to get a function list

Currently the store is fixed in UI, allowing ability to specify store in UI would allow to use function in store in UI. 
Although it doesn't serve the purpose well as I need to add the store url in UI every time

For now I need to make a complete separate project (a wrapper over faas-cli) to map of functions in store; which internally would use the faas-cli to deploy a function from store
  
Although If we Load store definition via Gateway,  it can be used to browse available functions which can be queried via both faas-cli and ui or any other platform written on top via API.

**Comment by rgee0** [comment]
@s8sg Could you more clearly express the use case, please.  It would help us understand what you are trying to do and whether there are more appropriate ways of doing it. 

Currently I'm left with the impression that automation is a factor here, which I wouldn't expect the UI to be impacted by.  I'm also unsure as to why creating a custom store would be necessary when a function stack can be deployed through the compose file or the CLI `stack.yml` file. 

The intention behind the UI store was to be a simple deployment mechanism for users - particularly new users - to be able to quickly deploy community curated and assured images from a central point and help bring themselves up to speed with the framework and the function(s).

**Comment by alexellis** [comment]
The function store supports an override for the store URL when using the CLI.

```
faas store --help
Allows browsing and deploying OpenFaaS functions from a store

Usage:
  faas-cli store [command]

Available Commands:
  deploy      Deploy OpenFaaS functions from a store
  inspect     Show details of OpenFaaS function from a store
  list        List available OpenFaaS functions in a store

Flags:
  -h, --help         help for store
  -u, --url string   Alternative Store URL starting with http(s):// (default "https://cdn.rawgit.com/openfaas/store/master/store.json")

Global Flags:
      --filter string   Wildcard to match with function names in YAML file
      --regex string    Regex to match with function names in YAML file
  -f, --yaml string     Path to YAML file describing function(s)

Use "faas-cli store [command] --help" for more information about a command.
```

I think while it's possible to extend the UI to load 1-many different function stores I'd like to understand the use-case behind that. The UI is an admin dashboard and not supposed to be the primary way to deploy functions for your team and business .

**Comment by alexellis** [comment]
@s8sg please see the updated comment here. Your thread seems to mix the CLI and the UI so perhaps  you could open a specific issue on the CLI repository instead?

Are you willing to do the work for your suggestion?

Also be aware that you can deploy a stack file via HTTP

```
faas cli deploy -f https://raw.githubusercontent.com/JockDaRock/nslookup_faas/master/nslookup.yml
```

**Comment by s8sg** [comment]
Hi @alexellis , I understand the idea behind not having a  store functionalities inside the gateway from @ericstoekl 's comment, which make complete sense. I also understand that current UI is `admin` portal.  I don't have much experience on frontend although I can definitely give it a try to add the store URL in the UI to browse and deploy functions. I'll update on the same soon


**OFF TOPIC:**

Initially When I imagine `stack.yml` its the function definitions. Like Docker-file in docker

By Function store I imagine it as a Docker registry. That means I can push and pull function to `store` just like we do in docker.
`faas-cli build   -f <yml> <funcname>`  
`faas-cli push  <funcname> -g <gateway>`  
which are already present.

But the only problem is, I can't just run the image as a function with only image, I need the yml file or pass the flags. i.e I can't do
`faas-cli deploy  <funcname> -g <gateway>`
Please correct me if its wrong. Cause we need to pass the `fprocess`env and other `env`. 

I thought store as the alternative. But from @ericstoekl comment I can understand the store purpose.  My idea was to have a function repository where I can login, browse and deploy ready to use functions which abstracts the docker repo. 
`faas-cli login <user> -g <gateway>`  
`faas-cli build  -f <yml > <funcname>`  
`faas-cli push  -f <yml > <funcname> -g <gateway>`  // will also store the function reference in gateway
`faas-cli list -a -g <gateway>`  // list all the available functions
`faas-cli list  -g <gateway>` // list only running functions in the gateway
`faas-cli pull  <funcame>  -g <gateway`   // Pull a ready to use function from a remote function repository
`faas-cli deploy  <funcname> -g <gateway>`  // pull if not present and deploy function 

** Note here functions doesn't require the stack.yml, but it itself is the ready to deploy entity. 

Also the docker-registry (remote/local) is abstracted by the gateway. 
Similar way we can have a `function-repository which abstracts the docker repo`

Currently I need to store at-least the `stack.yml`, which I was trying to avoid

Please suggest if there is any work around. 

I can reopen a separate issue in CLI if needed.

**Comment by BurtonR** [comment]
@s8sg  any luck with this?

**Comment by alexellis** [comment]
To support this we need a CORS rule for the UI, or a new server endpoint to serve the store data. The URL would have to be configured through the gateway at run-time. I could see this as being useful for an enterprise user who has all their images stored in a private repo, but we do support this via the CLI already.

Thoughts from all here?

Alex

**Comment by alexellis** [comment]
Derek close: inactivity

**Comment by DropBearHunta** [comment]
I would like to open this issue back up: as an enterprise user that uses a private repo for all of our images it would be nice to be able to customize the store url so that the UI is usable to deploy functions from our repo
